---
title: "Introduction to R"
author: "Maria Doyle"
date: "`r format(Sys.time(), '%d %B %Y')`"
output: 
  html_notebook:
    toc: yes
    toc_float: yes
    toc_depth: 2
subtitle: visualising RNA-seq data with ggplot2
---

## Acknowledgements
Some text remixed from [Data Carpentry](https://datacarpentry.org/R-ecology-lesson/04-visualization-ggplot2.html)

## Objectives
  
* Demonstrate how R and ggplot2 can be used to visualise RNA-Seq data
* Demonstrate basic R syntax (`<-`, `dim()`, `head()`, `View()`)
* Demonstrate how to create bar plots, density plots, box plots and violin plots with ggplot2

## Introduction

In this tutorial, we will learn some R through creating plots to visualise data from an RNA-seq experiment.


## Analysis

We will create some plots using published RNA-seq data from the Nature Cell Biology paper by [Fu et al. 2015] (https://www.ncbi.nlm.nih.gov/pubmed/25730472). This study examined expression in basal and luminal cells from mice at different stages (virgin, pregnant and lactating). There are 2 samples per group and 6 groups so 12 samples in total. We will use the raw counts here. These are the counts of reads for each gene for each sample. The higher the number of reads/counts the more the gene is expressed. Some of the plots are based on those in the [COMBINE RNA-seq R workshop](http://combine-australia.github.io/RNAseq-R/06-rnaseq-day1.html), except here we will create plots using **ggplot2*.

First we will load in the libraries we need. **`ggplot2`** is included in the **`tidyverse`** package. This is already installed for you on the server but if you need to install it on your own computer you can use `install.packages('tidyverse')`.

**The [tidyverse](https://www.tidyverse.org/) makes data science faster, easier and more fun.**

![www.tidyverse.org](images/tidyverse.png)


```{r, message=FALSE}
library(tidyverse)
```

The tidyverse is a collection of packages. We will introduce you to some of these packages in this course.

![Tidyverse packages](images/tidyverse_packages.jpeg)


The file we will use is tab-separated, so we will use the `read_tsv()` function from the tidyverse **`readr`** package to read it in. To see what the `read_tsv()` function (or any function in R) does, type a `?` before the name and the help will appear in the Help panel on the right in RStudio.

```{r, eval=FALSE}
?read_tsv
```

In R we use `<-` to assign values or data to variables. A variable is name we can use to store the data in R's memory. We can read in a file from a path on our computer on on the web. Note that we need to put quotes ("") around file paths.

```{r results = 'hide'}
counts <- read_tsv("/training/r-intro-tidyverse/data/counts.tsv.gz")
```


Let's take a look at the data. 

We can use `dim()` to see the number of rows and columns.

```{r}
dim(counts)
```

There are 326,148 rows and 5 columns.

We can take a look the first few lines with `head()`.

```{r}
head(counts)
```

We can look at the last few lines with 
```{r}
tail(counts)
```

Or we can see the whole file with `View()`.

```{r eval=FALSE}
View(counts)
```

## Plotting with **`ggplot2`**

**`ggplot2`** is a plotting package that makes it simple to create complex plots. We only need minimal changes if the underlying data change
or if we decide to change our plot type, for example, from a bar plot to a box plot. This helps in creating publication quality plots with minimal amounts of adjustments and tweaking.

**`ggplot2`** likes data in the 'long' format, also called 'tidy' format i.e., a column for every variable,
and a row for every observation. Well-structured data will save you lots of time
when making figures with **`ggplot2`**. We will discuss this more later in the course.

ggplot graphics are built step by step by adding new elements. Adding layers in
this fashion allows for extensive flexibility and customization of plots.

To build a ggplot, we will use the following basic template that can be used for different types of plots:

```{r}
ggplot(data=, mapping=aes()) + geom_ ()
```

There are different geom_ we can use to create different types of plot e.g. bar plot versus box plot, to see some of the geoms available see the ggplot2 reference here: https://ggplot2.tidyverse.org/reference/ and the ggplot2 cheat sheet here: https://www.rstudio.com/wp-content/uploads/2016/11/ggplot2-cheatsheet-2.1.pdf

An example scatter plot
To demonstrate how ggplot works we will use one of the built-in datasets that R has. The dataset is called women and it contains the average heights and weights for American women.

Take a look at the dataset.

```{r}
head(women)
```

Let's make a scatterplot. Note that to add a geom to the plot we use the `+` operator.

```{r}
ggplot(data=women, mapping=aes(x=height, y=weight)) + 
  geom_point()
```

Now we will explore the RNA-Seq data using different types of plots.

## Bar plots

How many counts do we have for each sample?
  
We can make bar plots to visualise this. To do this we will use `geom_bar()`. Let's take a look at the help for `geom_bar()`.

```{r eval=FALSE}
?geom_bar() 
```

We give `ggplot()` the data `counts` and in `aes()` we usually say which columns are the x and y axis values. However, in this case, note the part in the help that says "geom_bar() makes the height of the bar proportional to the number of cases in each group (or if the weight aesthetic is supplied, the sum of the weights)." Our input table contains the count for each gene for each sample but we want to plot the sum of counts for each sample. So we can specify `x=Sample` and use `weight=Counts` and ggplot will automatically sum the counts for each sample for us.

```{r}
ggplot(data=counts, mapping=aes(x=Sample, weight=Counts)) +
  geom_bar() 
```

We can colour by cell type (basal vs luminal) using the CellType column in our data. We simply need to add `fill=` into the `aes()` and specify the column in our data with the variable we want to colour.

```{r}
ggplot(data=counts, mapping=aes(x=Sample, weight=Counts, fill=CellType)) +
  geom_bar() 
```

A really nice feature about ggplot is that we can easily colour by another variable e.g. status (virgin vs pregnant vs lactating) by simply changing the column we give to `fill=`.

```{r}
ggplot(data=counts, mapping=aes(x=Sample, weight=Counts, fill=Status)) +
  geom_bar() 
```

## Density plots

In addition to viewing the number of counts per sample it is also useful to visualise the distribution of the counts. This helps us to compare the samples and check if any look unusual. We can visualise distributions with density plots. As count data is not normally distributed, to make it easier to visualise the distributions we log the counts. We'll plot the Sample on the X axis and log2 Counts on the y axis. As counts may be zero we add a small number to every count to avoid taking a log of zero. We'll add 0.25 to each count.

```{r}
ggplot(data=counts, mapping=aes(x=log2(Counts+0.25))) +
  geom_density()
```

We can colour by sample using `fill=`.

```{r}
ggplot(data=counts, mapping=aes(x=log2(Counts+0.25), fill=Sample)) +
  geom_density() 
```

If we would like to colour the lines in the plot we can use `colour=` instead of `fill=`. `fill=` is used to **fill** in areas in ggplot2 plots, whereas `colour=` is used to colour lines and points.


```{r}
ggplot(data=counts, mapping=aes(x=log2(Counts+0.25), colour=Sample)) +
  geom_density() 
```

The line at the bottom is present because geom_density plots a polygon shape. If we don't want that line we can use `geom_line()` e.g. https://stackoverflow.com/questions/49593634/unintended-line-across-x-axis-of-density-plot-r
```{r}
ggplot(data=counts, mapping=aes(x=log2(Counts+0.25), colour=Sample)) +
  geom_line(stat="density") 
```

## Box plots

We can also make box plots to visualise the distribution of counts for each sample.

```{r}
ggplot(data=counts, mapping=aes(x=Sample, y=log2(Counts))) +
  geom_boxplot()
```

#### Exercise

Colour the box plots by CellType and Status

```{r}
ggplot(data=counts, mapping=aes(x=Sample, y=log2(Counts), fill=CellType)) +
  geom_boxplot()
```


```{r}
ggplot(data=counts, mapping=aes(x=Sample, y=log2(Counts), fill=Status)) +
  geom_boxplot()
```

## Violin plots

Boxplots are useful summaries, but hide the *shape* of the distribution. For example, if the distribution is bimodal, we would not see it in a boxplot. An alternative to the boxplot is the violin plot, where the shape 
(of the density of points) is drawn.

#### Exercise

Make a violin plot instead of a box plot using geom_violin()

Solution

```{r}
ggplot(data=counts, mapping=aes(x=Sample, y=log2(Counts), fill=CellType)) +
  geom_violin()
```

```{r}
ggplot(data=counts, mapping=aes(x=Sample, y=log2(Counts), fill=Status)) +
  geom_violin()
```

## Saving plots

We can output plots to pdf using `pdf()` followed by `dev.off()`. We put our plot code after the call to `pdf()` and before closing the plot device with `dev.off()`.

```{r, eval=FALSE}
pdf("myplot.pdf")
ggplot(data=counts, mapping=aes(x=Sample, y=log2(Counts), fill=Status)) +
  geom_violin()
dev.off()
```

#### Exercise
  * Read in the normalised counts (from "/training/r-intro-tidyverse/data/normcounts.tsv.gz")
  * How many rows and columns are in the file
  * What is the first gene in the file and it's counts
  * What is the last gene in the file and it's counts
  * Make as many plots as you like from bar plots, density plots, boxplots and violin plots using the normalised counts (some starter code is below)
  * Colour by CellType and by Status
  * Do you notice any differences versus the unnormalised counts
  
```{r, eval=FALSE}
norm_counts <- read_tsv()
norm_counts
```


```{r, eval=FALSE}
ggplot(data= , mapping=aes(x= , weight= , fill= )) +
  geom_bar() 
```
 

```{r, eval=FALSE}
ggplot(data= , mapping=aes(x= , colour= )) +
  geom_density() 
```


```{r, eval=FALSE}
ggplot(data= , mapping=aes(x= , y= , fill= )) +
  geom_boxplot() 
```


```{r, eval=FALSE}
ggplot(data= , mapping=aes(x= , y= , fill= )) +
  geom_violin()
```

## Key Points
- We use `library()` to load packages we want to use. Note that they need to be already installed with e.g. `install.packages()`.
- We can read in tab-separated files with `read_tsv()`.
- We can check our data with `dim()`, `head()`, `tail()` and `View()`.
- **`ggplot2`** is a plotting package that makes it simple to create complex plots.
- ggplot2 plots have 3 components: data (dataset), mapping (variables/columns to plot) and geom (type of plot).
- The `+` sign is used to add geoms (and new layers as we will see later in the course). It must be placed at the end of the preceding line. If it is added at the beginning of the line, **`ggplot2`** will return an error message.
- ggplot2 plots can be easily coloured by variables/columns in the dataset. `fill=` can be used to colour areas and `colour=` to colour lines.
- A pdf of the plot can be generated with `pdf()` and `dev.off()`

## Further Reading
[An Introduction to ggplot](https://www.bioinformatics.babraham.ac.uk/training.html#ggplot) by Babraham Bioinformatics  
[Data Manipulation and Visualisation](http://bioinformatics-core-shared-training.github.io/r-intermediate/) by University of Cambridge 